test/rubygems/test_gem_platform.rb

```diff
@@ -261,6 +264,42 @@ def test_nil_cpu_arch_is_treated_as_universal
     assert((with_x86_arch === with_nil_arch), 'x86 =~ nil')
   end
 
+  def test_nil_version_is_treated_as_any_version
+    x86_darwin_8 = Gem::Platform.new 'i686-darwin8.0'
+    x86_darwin_nil = Gem::Platform.new 'i686-darwin'
+
+    assert((x86_darwin_8 === x86_darwin_nil), '8.0 =~ nil')
+    assert((x86_darwin_nil === x86_darwin_8), 'nil =~ 8.0')
+  end
+
+  def test_nil_version_is_stricter_for_linux_os
+    x86_linux = Gem::Platform.new 'i686-linux'
+    x86_linux_gnu = Gem::Platform.new 'i686-linux-gnu'
+    x86_linux_musl = Gem::Platform.new 'i686-linux-musl'
+    x86_linux_uclibc = Gem::Platform.new 'i686-linux-uclibc'
+
+    # a naked linux runtime is implicit gnu, as it represents the common glibc-linked runtime
+    assert(x86_linux === x86_linux_gnu, 'linux =~ linux-gnu')
+    assert(x86_linux_gnu === x86_linux, 'linux-gnu =~ linux')
+
+    # musl and explicit gnu should differ
+    # refute(x86_linux_gnu === x86_linux_musl, 'linux-gnu =~ linux-musl')
+    # refute(x86_linux_musl === x86_linux_gnu, 'linux-musl =~ linux-gnu')
+
+    # explicit libc differ
+    refute(x86_linux_uclibc === x86_linux_musl, 'linux-uclibc =~ linux-musl')
+    refute(x86_linux_musl === x86_linux_uclibc, 'linux-musl =~ linux-uclibc')
+
+    # musl host runtime accepts libc-generic or statically linked gems...
+    assert(x86_linux === x86_linux_musl, 'linux =~ linux-musl')
+    # ...but implicit gnu runtime generally does not accept musl-specific gems
+    refute(x86_linux_musl === x86_linux, 'linux-musl =~ linux')
+
+    # other libc are not glibc compatible
+    # refute(x86_linux === x86_linux_uclibc, 'linux =~ linux-uclibc')
```

Comment:
The case of these non-musl, non-gnu libc that are known not to be compatible at all with glibc binaries (so zero chance of `linux` working) remains unsolved here. Anecdotal I know, but one day musl was anecdotal as well.

Reply:
I fixed this one too! Never seen these in the wild, but if they're not compatible it seems good to make sure we never match "standard" linux.